dbms/src/Storages/MergeTree/DiskSpaceMonitor.cpp

```diff
@@ -1,10 +1,223 @@
+            path = config.getString(disk_config_prefix + ".path");
+
+        if (disk_name == default_disk_name)
+        {
+            has_default_disk = true;
+            if (!path.empty())
+                throw Exception("It is not possible to specify default disk path", ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
+            disks.emplace(disk_name, std::make_shared<const Disk>(disk_name, default_path, keep_free_space_bytes));
+        }
+        else
+        {
+            if (path.empty())
+                throw Exception("Disk path can not be empty. Disk " + disk_name, ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
+            disks.emplace(disk_name, std::make_shared<const Disk>(disk_name, path, keep_free_space_bytes));
+        }
+    }
+    if (!has_default_disk)
+        disks.emplace(default_disk_name, std::make_shared<const Disk>(default_disk_name, default_path, 0));
+}
+
+const DiskPtr & DiskSelector::operator[](const String & name) const
+{
+    auto it = disks.find(name);
+    if (it == disks.end())
+        throw Exception("Unknown disk " + name, ErrorCodes::UNKNOWN_DISK);
+    return it->second;
+}
+
+bool DiskSelector::has(const String & name) const
+{
+    auto it = disks.find(name);
+    return it != disks.end();
+}
+
+void DiskSelector::add(const DiskPtr & disk)
+{
+    disks.emplace(disk->getName(), disk);
+}
+
+Schema::Volume::Volume(const Poco::Util::AbstractConfiguration & config, const std::string & config_prefix, const DiskSelector & disk_selector)
+{
+    Poco::Util::AbstractConfiguration::Keys keys;
+    config.keys(config_prefix, keys);
+
+    for (const auto & name : keys)
+    {
+        if (startsWith(name, "disk"))
+        {
+            auto disk_name = config.getString(config_prefix + "." + name);
+            disks.push_back(disk_selector[disk_name]);
+        }
+    }
+
+    if (disks.empty())
+        throw Exception("Volume must contain at least one disk", ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
+
+    auto has_max_bytes = config.has(config_prefix + ".max_data_part_size_bytes");
+    auto has_max_ratio = config.has(config_prefix + ".max_data_part_size_ratio");
+    if (has_max_bytes && has_max_ratio)
+    {
+        throw Exception("Only one of 'max_data_part_size_bytes' and 'max_data_part_size_ratio' should be specified",
+                        ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
+    }
+
+    if (has_max_bytes)
+    {
+        max_data_part_size = config.getUInt64(config_prefix + ".max_data_part_size_bytes");
+    }
+    else if (has_max_ratio)
+    {
+        auto ratio = config.getDouble(config_prefix + ".max_data_part_size_bytes");
+        if (ratio < 0 and ratio > 1)
+        {
+            throw Exception("'max_data_part_size_bytes' have to be between 0 and 1",
+                            ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
+        }
+        UInt64 sum_size = 0;
+        for (const auto & disk : disks)
+            sum_size += disk->getTotalSpace();
```

Comment:
I bit controversial. Lets say you have 2 equal disks & set up ratio to 0.6. That part will never fit any of the disks. May be it should be calculated as related to maximum / minimum disk size from volume?

Reply:
Maybe median or average of disk sizes?

Reply:
IMHO it should not change significantly when new disks are added. I.e. if before adding new disk max calculated part size was 100Mb, i would expect that allowed partsize for that volume still remains unchanged even after adding new bigger disk. That sounds like min * ratio.

Reply:
For me that usecase (with ratio) is a bit non clear. May be we can lease as is (with sum) but introduce some extra check which will guarantee that each disk in volume can store at least few parts of that size?

Reply:
> For me that usecase (with ratio) is a bit non clear. May be we can leave it as is (with sum) but introduce some extra check which will guarantee that each disk in volume can store at least few parts of that size?

I think we should use average. Ratio should be from 0 to inf. It would not create reservation on disk if part greater than disk can store (current available space).